src/daemon/daemon.cpp

```diff
@@ -678,26 +775,45 @@ mp::Daemon::Daemon(std::unique_ptr<const DaemonConfig> the_config)
 {
     connect_rpc(daemon_rpc, *this);
     std::vector<std::string> invalid_specs;
-    bool mac_addr_missing{false};
+
     for (auto& entry : vm_instance_specs)
     {
         const auto& name = entry.first;
         auto& spec = entry.second;
 
+        // Store here the MAC addresses we need add to the daemon's set. But we'll add them only if this instance
+        // is not invalid.
+        std::unordered_set<std::string> instance_mac_addresses;
+
         if (!config->vault->has_record_for(name))
         {
             invalid_specs.push_back(name);
             continue;
         }
 
-        auto mac_addr = spec.mac_addr;
-        if (mac_addr.empty())
+        // Check that all the interfaces in the instance have different MAC address, and that they were not used in
+        // the other instances. String validity was already checked in load_db().
+        auto new_macs = mac_set_from(spec);
+
+        if (new_macs.size() <= spec.extra_interfaces.size())
+        {
+            // There is at least one repeated address in new_macs.
+            mpl::log(mpl::Level::warning, category, fmt::format("{} has repeated MAC addresses", name));
+            invalid_specs.push_back(name);
+            continue;
+        }
+        else if (new_macs.merge(allocated_mac_addrs); !allocated_mac_addrs.empty())
+        {
+            // There is a repeated address from a previous instance.
+            mpl::log(mpl::Level::warning, category, fmt::format("{} has repeated MAC addresses", name));
+            invalid_specs.push_back(name);
+            continue;
+        }
+        else
         {
-            mac_addr = mp::utils::generate_mac_address();
-            vm_instance_specs[name].mac_addr = mac_addr;
-            mac_addr_missing = true;
+            // If there are no repetitions, add the new macs to the daemon's list.
+            allocated_mac_addrs = std::move(new_macs);
         }
```

Comment:
The body of the two "ifs" is exactly identical, so they can be combined. Using the comma operator, we can still avoid the merge if there is internal repetition. Finally, no need for the else since you `continue` in the if.

Github doesn't allow me to "suggest" on unchanged code, so issuing a small PR instead.